### PR TITLE
FallDamage binder

### PR DESCRIPTION
**Pull request by planetme**
Created handler for fall damage and added a config to scale fall damage and height threshold for damage.

### CODE REVIEW

**Comment by Hexcede** [comment]
I'd probably rename the `Fall` config to `FallDamage` to match the binder

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 1]
```diff
@@ -0,0 +1,50 @@
+local PlayerService = game:GetService("Players")
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+local Config = require(ReplicatedStorage.Common.Config.FallDamage)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: Rename this to FallDamageConfig

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+FallDamage.__index = FallDamage
+
+function FallDamage:_detectFallHeight(character: Model, humanoid: Humanoid, signal: Signal)
+    local initialHeight = nil
+
+    humanoid.FreeFalling:Connect(function(active)
+        if active then
+            initialHeight = character.HumanoidRootPart.Position.Y
+        else
+			signal:Fire(initialHeight - character.HumanoidRootPart.Position.Y)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: What happened here?

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        end
+    end)
+end
+
+function FallDamage:_updateHealth(fallHeight, humanoid: Humanoid)
+    if fallHeight < Config.threshold then return else
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: Turn this into a guard clause.

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+end
+
+function FallDamage.new(character: Model)
+	local self = setmetatable({}, FallDamage)
+	
+    self._character = character
+    self._humanoid = self._character:WaitForChild("Humanoid")
+    self._trove = Trove.new()
+    self._fallSignal = self._trove:Construct(Signal)
+    self.Health = self._humanoid.Health
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: Why is health public?

File: src/ReplicatedStorage/Common/Config/FallDamage.lua [hunk 5]
```diff
@@ -0,0 +1,4 @@
+return {
```
Comment (4812571) [inline src/ReplicatedStorage/Common/Config/FallDamage.lua]: Explicitly type local FallDamageConfig = {

}

File: .vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
+    "robloxLsp.diagnostics.disable": [
+        "undefined-type"
```
Comment (4812571) [inline .vscode/settings.json]: Why did you disable this LSP setting?

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 7]
```diff
@@ -0,0 +1,50 @@
+local PlayerService = game:GetService("Players")
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+local Config = require(ReplicatedStorage.Common.Config.FallDamage)
+local Signal = require(ReplicatedStorage.Packages.Signal)
+local Trove = require(ReplicatedStorage.Packages.Trove)
+
+local FallDamage = {}
+FallDamage.__index = FallDamage
+
+function FallDamage:_detectFallHeight(character: Model, humanoid: Humanoid, signal: Signal)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: Is this typedef for ``Signal`` correct?

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        humanoid.Health -= damage
+        return damage
+    end
+end
+
+function FallDamage.new(character: Model)
+	local self = setmetatable({}, FallDamage)
+	
+    self._character = character
+    self._humanoid = self._character:WaitForChild("Humanoid")
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: Is it necessary for these variables to be in self?

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    self._trove = Trove.new()
+    self._fallSignal = self._trove:Construct(Signal)
+    self.Health = self._humanoid.Health
+
+    self:_detectFallHeight(self._character, self._humanoid, self._fallSignal)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: Why are you passing the the fall Signal if it's in self already?

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+local Signal = require(ReplicatedStorage.Packages.Signal)
+local Trove = require(ReplicatedStorage.Packages.Trove)
+
+local FallDamage = {}
+FallDamage.__index = FallDamage
+
+function FallDamage:_detectFallHeight(character: Model, humanoid: Humanoid, signal: Signal)
+    local initialHeight = nil
+
+    humanoid.FreeFalling:Connect(function(active)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: This is a memory leak! Use Trove connect!

**Review by 4812571 — CHANGES_REQUESTED** [review]

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+function FallDamage:_updateHealth(fallHeight, humanoid: Humanoid)
+    if fallHeight < Config.threshold then return else
+        local damage = (fallHeight - Config.threshold) * Config.scale
+        humanoid.Health -= damage
+        return damage
+    end
+end
+
+function FallDamage.new(character: Model)
+	local self = setmetatable({}, FallDamage)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: What happened here?

**Review by 4812571 — CHANGES_REQUESTED** [review]

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+			self._fallSignal:Fire(initialHeight - character.HumanoidRootPart.Position.Y)
+        end
+    end))
+end
+
+function FallDamage:_updateHealth(fallHeight, humanoid: Humanoid)
+    if fallHeight < FallDamageConfig.threshold then return end
+
+    local damage = (fallHeight - FallDamageConfig.threshold) * FallDamageConfig.scale
+    humanoid.Health -= damage
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: This should use humanoid:TakeDamage

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+FallDamage.__index = FallDamage
+
+function FallDamage:_detectFallHeight(character: Model, humanoid: Humanoid)
+    local initialHeight = nil
+
+    self._trove:Connect(humanoid.FreeFalling:Connect(function(active)
+        if active then
+            initialHeight = character.HumanoidRootPart.Position.Y
+        else
+			self._fallSignal:Fire(initialHeight - character.HumanoidRootPart.Position.Y)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: whitespace issue

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+function FallDamage.new(character: Model)
+    local humanoid = character:WaitForChild("Humanoid")
+
+    local self = setmetatable({}, FallDamage)
+
+    self._trove = Trove.new()
+    self._fallSignal = self._trove:Construct(Signal)
+    self:_detectFallHeight(character, humanoid)
+    self._trove:Connect(self._fallSignal, function(fallHeight) self:_updateHealth(fallHeight, humanoid) end)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: self:_updateHealth can be moved into the detection code and thus the signal can be removed. Additionally, 'detectFallHeight' implies that the check happens once, try renamining it to something like 'trackCharacterFall' or something more descriptive and correct.

**Review by 4812571 — CHANGES_REQUESTED** [review]

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+function FallDamage:_trackCharacterFall(character: Model, humanoid: Humanoid)
+    local initialHeight = nil
+    local fallHeight = nil
+
+    self._trove:Connect(humanoid.FreeFalling, function(active)
+        if active then
+            initialHeight = character.HumanoidRootPart.Position.Y
+        else
+			fallHeight = initialHeight - character.HumanoidRootPart.Position.Y
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua:16]: Something wacky is happening here with whitespace. Try converting spaces to tabs first.

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+            if fallHeight < FallDamageConfig.threshold then return end
+
+            local damage = (fallHeight - FallDamageConfig.threshold) * FallDamageConfig.scale
+            humanoid:TakeDamage(damage)
+
+            return
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: Move this into FallDamage:_updateHealth

**Review by 4812571 — CHANGES_REQUESTED** [review]

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+        if active then
+            initialHeight = character.HumanoidRootPart.Position.Y
+        else
+            fallHeight = initialHeight - character.HumanoidRootPart.Position.Y
+            if fallHeight < FallDamageConfig.Threshold then return end
+
+            local damage = (fallHeight - FallDamageConfig.Threshold) * FallDamageConfig.Scale
+            humanoid:TakeDamage(damage)
+
+            return
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua]: This return can be eliminated

File: src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    end)
+end
+
+function FallDamage.new(character: Model)
+    local humanoid = character:WaitForChild("Humanoid")
+
+    local self = setmetatable({}, FallDamage)
+
+    self._trove = Trove.new()
+	self:_trackCharacterFall(character, humanoid)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Misc/FallDamage.lua:31]: Whitespace issue, try converting spaces to tabs

**Review by 4812571 — CHANGES_REQUESTED** [review]

**Comment by nottirb** [comment]
I would recommend changing spaces to tabs. You have a mix of both.

File: src/ServerScriptService/InstanceComponents/Misc/Character.lua [hunk 19]
```diff
@@ -23,6 +23,7 @@ function Character.new(instance: Instance)
 
 	-- Add CameraLook tag to the character
 	CollectionService:AddTag(instance, "CameraLook")
+	CollectionService:AddTag(instance, "FallDamage")
```
Comment (nottirb) [inline src/ServerScriptService/InstanceComponents/Misc/Character.lua]: The fall damage tag doesn't make sense to be under the "CameraLook tag" comment block. I would recommend changing the comment to read "Add character tags" or something of the sort.

Comment (nottirb) [reply src/ServerScriptService/InstanceComponents/Misc/Character.lua]: ``Add tags to the character`` might be better wording.

**Review by 4812571 — APPROVED** [review]

**Comment by 4812571** [comment]
LGTM!

**Pull request merged**